### ISSUE TITLE
using IAM for Service Accounts does not work

### ISSUE THREAD

**Issue by arohner** [kubernetes-sigs/aws-load-balancer-controller#1668]
Using version 2.0.1 on EKS. Using IAM Roles for Service Accounts, described here: https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html

There appears to be a bug with the way the container handles AWS credentials.

I created an IAM policy based on https://raw.githubusercontent.com/kubernetes-sigs/aws-alb-ingress-controller/main/docs/install/iam_policy.json

If I attach the policy to an IAM role, and attach the role to the ServiceAccount, cloudtrail shows the container successfully calls sts:AssumeRoleWithWebIdentity, but then performs no other actions. The log is filled with

```
{"level":"error","ts":1605710399.3141577,"logger":"controller","msg":"Reconciler error","controller":"ingress","name":"beta","namespace":"default","error":"UnauthorizedOperation: You are not authorized to perform this operation.\n\tstatus code: 403, request id: cf2913f4-20cd-43fd-9e9a-46117954ae84"}
```

If instead I attach the exact same policy to the EKS worker node role, everything works fine. 

If I hack the controller deployment pod definition to run our application image rather than the `load-balancer-controller` image, I'm able to `aws sts get-caller-identity` and `aws elbv2 describe-target-groups` just fine. To me, this indicates that the role is configured properly and attaching to the serviceaccount properly, but the official `load-balancer-controller` image is not using the webidentitytoken properly.

**Comment by kishorj** [comment]
Could you verify if your service account has the following annotatoin?
`eks.amazonaws.com/role-arn:arn:aws:iam::...`

Also check if you hit the eksctl issue https://github.com/weaveworks/eksctl/issues/2665.

**Comment by M00nF1sh** [comment]
@arohner 
Is this issue still exists? if so, you can send us your cluster-arn

**Comment by arohner** [comment]
> Could you verify if your service account has the following annotatoin?
eks.amazonaws.com/role-arn:arn:aws:iam::...

Yes, it does. I can tell the role works properly by starting the controller deployment with my own image, `kubectl exec`ing into it, and running `aws sts get-caller-identity` and `aws elbv2 describe-target-groups`, and seeing the calls work as expected

**Comment by arohner** [comment]
Yes, the issue still exists. The cluster arn is `arn:aws:eks:eu-west-2:623060147434:cluster/griffin-eks`

**Comment by kroka** [comment]
We are affected as well with _aws-load-balancer-controller:v2.0.1_ and Kubernetes v1.18.9 on EKS with _Fargate_.

The deployment succeeds, but the pod logs are full of 

```
UnauthorizedOperation: You are not authorized to perform this operation. Encoded authorization failure message: SQBWy..
```
where the decoded failure message shows
```
{
  "allowed": false,
  "explicitDeny": false,
  "matchedStatements": {
    "items": []
  },
  "failures": {
    "items": []
  },
  "context": {
    "principal": {
      "id": "***:***",
      "arn": "arn:aws:sts::***:assumed-role/eksctl-***-addon-iamse-Role1-***/***"
    },
    "action": "ec2:CreateTags",
    "resource": "arn:aws:ec2:eu-west-1:***:security-group/sg-***",
    "conditions": {
...
```
As far as I understand, the pod should be able to execute `ec2:CreateTags` thanks to the policy.


> Could you verify if your service account has the following annotatoin?

Looks good to me, the service account aws-load-balancer-controller has the annotation `arn:aws:iam::***:role/eksctl-***`

The pod definition does not look that bad either:
```
serviceAccountName: aws-load-balancer-controller
serviceAccount: aws-load-balancer-controller
```

**Comment by kroka** [comment]
In our case it was my bad. We had the _ALB Ingress Controller_ installed initially and wanted to start from scratch instead of upgrading to the _AWS Load Balancer Controller_. Hence, we've uninstalled all former "components", roles, bindings etc. but missed the security groups 

managed LoadBalancer securityGroup by ALB Ingress Controller

**Comment by kishorj** [comment]
@arohner, apologies for late response, I remember I had drafted the response but somehow I missed posting it.

I did review your cluster's configuration, and the service account annotation appeared correct. However, the aws-load-balancer-controller pod did not have the aws-iam-token configuration when I checked. It could be possible that the controller pod got created before the service account got configured with the  `eks.amazonaws.com/role-arn` annotation. You mention that your custom application works just fine -  this could be after the serviceaccount had the correct configuration.

If IRSA is setup correctly, your pod spec should have the following volume mount and volume injected - 

```
    - mountPath: /var/run/secrets/eks.amazonaws.com/serviceaccount
      name: aws-iam-token
      readOnly: true
```
 
```
  volumes:
  - name: aws-iam-token
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          audience: sts.amazonaws.com
          expirationSeconds: 86400
          path: token
```
